### PR TITLE
Write CRAM on Spark.

### PR DESCRIPTION
**Pull request by cmnbroad**
https://github.com/broadinstitute/gatk/issues/1270. Depends on https://github.com/broadinstitute/gatk/pull/1469, which should be reviewed first and then I can remove the first commit from this branch.

### CODE REVIEW

**Comment by cmnbroad** [comment]
@droazen This is ready whenever you are. In addition to writing CRAM on Spark, there are a couple of changed for https://github.com/broadinstitute/gatk/issues/1087, though we can't completely fix that yet due to the NM/MD tag issue.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 1]
```diff
@@ -83,24 +88,58 @@ public SparkHeaderlessBAMOutputFormat() {
         }
     }
 
+    // Output format class for writing CRAM files through saveAsNewAPIHadoopFile. Must be public.
+    public static class SparkCRAMOutputFormat extends KeyIgnoringCRAMOutputFormat<NullWritable> {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: This class is identical to `SparkBAMOutputFormat` -- is it only a separate class because `setHeader()` is static and we might want a different header for the two output formats, or is there some deeper reason?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 2]
```diff
@@ -125,9 +167,9 @@ public static void writeReads(
 
         if (format == ReadsWriteFormat.SINGLE) {
             header.setSortOrder(SAMFileHeader.SortOrder.coordinate);
-            writeReadsSingle(ctx, absoluteOutputFile, samReads, header, numReducers);
+            writeReadsSingle(ctx, absoluteOutputFile, absoluteReferenceFile, samOutputFormat, samReads, header, numReducers, conf);
         } else if (format == ReadsWriteFormat.SHARDED) {
-            saveAsShardedHadoopFiles(ctx, absoluteOutputFile, samReads, header, true);
+            saveAsShardedHadoopFiles(ctx, absoluteOutputFile, absoluteReferenceFile, samOutputFormat, samReads, header, true, conf);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Does `conf` need to be a separate parameter to these methods, given that it can be derived from `ctx`?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 3]
```diff
@@ -156,40 +198,53 @@ private static void writeReadsADAM(
...
+            SparkCRAMOutputFormat.setHeader(header);
+        } else {
+            SparkBAMOutputFormat.setHeader(header);
+        }
 
         final Broadcast<SAMFileHeader> headerBroadcast = ctx.broadcast(header);
 
-        // SparkBAMOutputFormat is a static class, so we need to copy the header to each worker then call
-        final JavaRDD<SAMRecord> readsRDD = setHeaderForEachPartition(reads, headerBroadcast);
+        // SparkBAM/CRAMOutputFormat are a static classes, so we need to copy the header to each worker then call
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: `a static classes` -> `static classes`

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 4]
```diff
@@ -156,40 +198,53 @@ private static void writeReadsADAM(
...
-            final SAMFileHeader header, final int numReducers) throws IOException {
+            final JavaSparkContext ctx, final String outputFile, String referenceFile, SAMFormat samOutputFormat, final JavaRDD<SAMRecord> reads,
+            final SAMFileHeader header, final int numReducers, Configuration conf) throws IOException {
 
         final JavaRDD<SAMRecord> sortedReads = sortReads(reads, header, numReducers);
-        saveAsShardedHadoopFiles(ctx, outputFile, sortedReads, header, false);
-        mergeHeaderlessBamShards(outputFile, header);
-
+        saveAsShardedHadoopFiles(ctx, outputFile, referenceFile, samOutputFormat, sortedReads,  header, false, conf);
+        mergeHeaderlessBamShards(ctx, outputFile, samOutputFormat, header);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Merging headerless CRAM shards is working?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 5]
```diff
@@ -156,40 +198,53 @@ private static void writeReadsADAM(
...
-        // SparkBAMOutputFormat is a static class, so we need to copy the header to each worker then call
-        final JavaRDD<SAMRecord> readsRDD = setHeaderForEachPartition(reads, headerBroadcast);
+        // SparkBAM/CRAMOutputFormat are a static classes, so we need to copy the header to each worker then call
+        final JavaRDD<SAMRecord> readsRDD = setHeaderForEachPartition(reads, samOutputFormat, headerBroadcast);
 
         // The expected format for writing is JavaPairRDD where the key is ignored and the value is SAMRecordWritable.
         final JavaPairRDD<SAMRecord, SAMRecordWritable> rddSamRecordWriteable = pairReadsWithSAMRecordWritables(headerBroadcast, readsRDD);
 
-        rddSamRecordWriteable.saveAsNewAPIHadoopFile(outputFile, SAMRecord.class, SAMRecordWritable.class, getOutputFormat(writeHeader));
+        rddSamRecordWriteable.saveAsNewAPIHadoopFile(outputFile, SAMRecord.class, SAMRecordWritable.class, getOutputFormat(samOutputFormat, writeHeader), conf);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: `SAMRecordWritable` is used for `CRAMOutputFormat` as well?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 6]
```diff
@@ -248,9 +308,14 @@ private static void mergeHeaderlessBamShards(String outputFile, SAMFileHeader he
         fs.delete(outputPath, true);
 
         try (final OutputStream out = fs.create(outputPath)) {
-            new SAMOutputPreparer().prepareForRecords(out, SAMFormat.BAM, header); // write the header
+            new SAMOutputPreparer().prepareForRecords(out, samOutputFormat, header); // write the header
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Is `header` here guaranteed to be the same as the one set on the corresponding `OutputFormat` class?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 7]
```diff
@@ -248,9 +308,14 @@ private static void mergeHeaderlessBamShards(String outputFile, SAMFileHeader he
...
             mergeInto(out, tmpPath, conf);
-            out.write(BlockCompressedStreamConstants.EMPTY_GZIP_BLOCK); // add the BGZF terminator
+            if (SAMFormat.CRAM == samOutputFormat) {
+                CramIO.issueEOF(CramVersions.DEFAULT_CRAM_VERSION, out); // terminate with CRAM EOF container
+            }
+            else {
+                out.write(BlockCompressedStreamConstants.EMPTY_GZIP_BLOCK); // add the BGZF terminator
+            }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Extract a `writeTerminatorBlock()` method here.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 8]
```diff
@@ -216,7 +218,12 @@ private void setHadoopBAMConfigurationProperties(final String inputName, final S
         final Configuration conf = ctx.hadoopConfiguration();
         conf.set(SAMHeaderReader.VALIDATION_STRINGENCY_PROPERTY, validationStringency.name());
 
-        if (!IOUtils.isCramFileName(inputName)) { // only set the reference for CRAM input
+        // If the input we're reading is sharded output, filePath is the name of a folder that contains
+        // the individual shard parts, and the original file extension (i.e., ".cram") is embedded in
+        // the middle of the name, so we need to scan the whole name to detect CRAM input
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Is it at a fixed position in the name (eg., second-to-last token if you split on `.`), or do we have to scan the entire String (which could produce false-positives for sharded cram input)

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java [hunk 9]
```diff
@@ -19,70 +20,75 @@ public String getTestedClassName() {
...
-        };
-        Assert.assertEquals(runCommandLine(args), null);
-        SamAssertionUtils.assertSamsEqual(outFile, ORIG_BAM);
-    }
-
-    @Test(dataProvider="testingDataCRAM")
-    public void testFileToFileCRAM(String fileIn, String extOut, String reference) throws Exception {
+    public void testFileToFile(String fileIn, String extOut, String reference) throws Exception {
         String samFile= fileIn;
         final File outFile = File.createTempFile(samFile + ".", extOut);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: Use `BaseTest.createTempFile()` instead to ensure indices, etc., are cleaned up

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java [hunk 10]
```diff
@@ -19,70 +20,75 @@ public String getTestedClassName() {
...
+                // TODO: We still need lenient comparison due to https://github.com/samtools/htsjdk/issues/455
+                // (cram->bam conversion strips out NM/MD tags)
+                SamAssertionUtils.assertSamsEqualLenient(ORIG_BAM, outFile, refFile);
+            }
+            else {
+                SamAssertionUtils.assertSamsEqual(ORIG_BAM, outFile, refFile);
+            }
         } else {
-            SamAssertionUtils.assertSamsEqual(outFile, ORIG_BAM, refFile);
+            SamAssertionUtils.assertSamsEqual(ORIG_BAM, outFile);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: Order is wrong here -- should be `actual, expected`

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java [hunk 11]
```diff
@@ -19,70 +20,75 @@ public String getTestedClassName() {
...
 
-        //TODO: Make this comparison non-lenient in all cases: https://github.com/broadinstitute/gatk/issues/1087
-        if (extOut.equals(".cram")){
-            SamAssertionUtils.assertSamsEqualLenient(outFile, ORIG_BAM, refFile);
+        if (fileIn.endsWith(CramIO.CRAM_FILE_EXTENSION) || extOut.equals(CramIO.CRAM_FILE_EXTENSION)) {
+            final File refFile = new File(TEST_DATA_DIR, reference);
+            if (!fileIn.endsWith(CramIO.CRAM_FILE_EXTENSION)) {
+                // TODO: We still need lenient comparison due to https://github.com/samtools/htsjdk/issues/455
+                // (cram->bam conversion strips out NM/MD tags)
+                SamAssertionUtils.assertSamsEqualLenient(ORIG_BAM, outFile, refFile);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: `actual, expected`

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java [hunk 12]
```diff
@@ -19,70 +20,75 @@ public String getTestedClassName() {
...
+            }
+            else {
+                SamAssertionUtils.assertSamsEqual(ORIG_BAM, outFile, refFile);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: `actual, expected`

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java [hunk 13]
```diff
@@ -19,70 +20,75 @@ public String getTestedClassName() {
...
         runCommandLine(args);
 
-        //TODO: Make this comparison non-lenient in all cases: https://github.com/broadinstitute/gatk/issues/1087
-        if (extOut.equals(".cram")){
-            SamAssertionUtils.assertSamsEqualLenient(outFile, ORIG_BAM, refFile);
+        if (fileIn.endsWith(CramIO.CRAM_FILE_EXTENSION) || extOut.equals(CramIO.CRAM_FILE_EXTENSION)) {
+            final File refFile = new File(TEST_DATA_DIR, reference);
+            if (!fileIn.endsWith(CramIO.CRAM_FILE_EXTENSION)) {
+                // TODO: We still need lenient comparison due to https://github.com/samtools/htsjdk/issues/455
+                // (cram->bam conversion strips out NM/MD tags)
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: Does `bam->cram` not strip out these tags as well?

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java [hunk 14]
```diff
@@ -19,70 +20,75 @@ public String getTestedClassName() {
...
+                {"print_reads.sorted.queryname.bam", ".bam", null},
+                {"print_reads.sorted.queryname.bam", ".cram", "print_reads.fasta"},
                 {"print_reads.sorted.queryname.cram", ".sam", "print_reads.fasta"},
                 {"print_reads.sorted.queryname.cram", ".bam", "print_reads.fasta"},
-
-                //comparing queryname sorted cram files blows up: https://github.com/broadinstitute/gatk/issues/1271
-//                {"print_reads.sorted.queryname.cram", ".cram", "print_reads.fasta"},
-//                {"print_reads.sorted.queryname.sam", ".cram", "print_reads.fasta"},
-//                {"print_reads.sorted.queryname.bam", ".cram", "print_reads.fasta"}
+                {"print_reads.sorted.queryname.cram", ".cram", "print_reads.fasta"}
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: Was this cram file created with the new cram-writing code in htsjdk that won't create new containers inappropriately?

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java [hunk 15]
```diff
@@ -25,17 +26,15 @@ public String getTestedClassName() {
...
-//                {"print_reads.sorted.cram", ".cram", "print_reads.fasta"},
-//                {"print_reads.sorted.sam", ".cram", "print_reads.fasta"},
-//                {"print_reads.sorted.bam", ".cram", "print_reads.fasta"}
+                {"print_reads.sorted.sam", ".cram", "print_reads.fasta"},
+                {"print_reads.sorted.bam", ".cram", "print_reads.fasta"},
+                {"print_reads.sorted.cram", ".sam", "print_reads.fasta"},
+                {"print_reads.sorted.cram", ".bam", "print_reads.fasta"},
+                {"print_reads.sorted.cram", ".cram", "print_reads.fasta"}
         };
     }
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java]: Use `BaseTest.createTempFile()` in `testFileToFile()` below

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java [hunk 16]
```diff
@@ -61,43 +60,14 @@ public void testFileToFile(String fileIn, String extOut, String reference) throw
...
-        if (reference == null) {
-            refFile = null;
-            args = new String[]{
-                    "--input", originalFile.getAbsolutePath(),
-                    "--output", outFile.getAbsolutePath(),
-            };
+        //TODO - converting a sam/bam to cram strips out the NM/MD tags so we still
+        // need to use lenient comparison for those
+        if (extOut.equals(CramIO.CRAM_FILE_EXTENSION) && !fileIn.endsWith(CramIO.CRAM_FILE_EXTENSION)) {
+            SamAssertionUtils.assertSamsEqualLenient(originalFile, outFile, refFile);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java]: `actual, expected`

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java [hunk 17]
```diff
@@ -61,43 +60,14 @@ public void testFileToFile(String fileIn, String extOut, String reference) throw
...
         } else {
-            refFile = new File(TEST_DATA_DIR, reference);
-            args = new String[]{
-                    "--input", originalFile.getAbsolutePath(),
-                    "--output", outFile.getAbsolutePath(),
-                    "-R", refFile.getAbsolutePath()
-            };
+            SamAssertionUtils.assertSamsEqual(originalFile, outFile, refFile);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java]: `actual, expected`

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java [hunk 18]
```diff
@@ -61,43 +60,14 @@ public void testFileToFile(String fileIn, String extOut, String reference) throw
...
-        final File refFile;
-        final String[] args;
-        if (reference == null) {
-            refFile = null;
-            args = new String[]{
-                    "--input", originalFile.getAbsolutePath(),
-                    "--output", outFile.getAbsolutePath(),
-            };
+        //TODO - converting a sam/bam to cram strips out the NM/MD tags so we still
+        // need to use lenient comparison for those
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java]: Link to relevant htsjdk ticket here

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java [hunk 19]
```diff
@@ -157,11 +127,9 @@ public void testSequenceDictionaryValidation() throws Exception {
...
                 {"print_reads.sorted.queryname.cram", ".sam", "print_reads.fasta"},
                 {"print_reads.sorted.queryname.cram", ".bam", "print_reads.fasta"},
-
-                //comparing queryname sorted cram files blows up: https://github.com/broadinstitute/gatk/issues/1271
-//                {"print_reads.sorted.queryname.cram", ".cram", "print_reads.fasta"},
-//                {"print_reads.sorted.queryname.sam", ".cram", "print_reads.fasta"},
-//                {"print_reads.sorted.queryname.bam", ".cram", "print_reads.fasta"}
+                {"print_reads.sorted.queryname.cram", ".cram", "print_reads.fasta"},
+                {"print_reads.sorted.queryname.sam", ".cram", "print_reads.fasta"},
+                {"print_reads.sorted.queryname.bam", ".cram", "print_reads.fasta"}
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java]: Do these test cases pass with a non-lenient assert despite the tag stripping?

**Comment by droazen** [comment]
Review complete -- back to @cmnbroad. Merge after addressing comments.

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java (hunk 14 above)
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: If not, might want to add a name-sorted cram produced by a recent htsjdk as a test case, since internally the format is very different than it was before.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java (hunk 1 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Its only because the two classes have different base classes that live in Hadoop-BAM. I don't see any useful way to consolidate them here, though going forward we may want to change AnySAMOutputFormat in Hadoop-BAM to delegate based on file extension similar to the way AnySAMInputFormat does.

Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Created https://github.com/HadoopGenomics/Hadoop-BAM/issues/72.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java (hunk 2 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: No - looks like it can  be removed.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java (hunk 5 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: AFAIK, the "writable" part of SAMRecordWritable is used for record serialization during shuffle serialization, but the file output format is what dictates the file format (and record writer) for file save, so I think this gets the behavior we want.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java (hunk 4 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Yes.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java (hunk 6 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: mergeHeaderlessBamShards is only called from writeReadsSingle, which passes the same header through to both the output class and the merging code. This reminds me, though, that we should change Hadoop-BAM (as well as GATK which we've previously discussed) to use setHeaderStrict.

Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Created https://github.com/HadoopGenomics/Hadoop-BAM/issues/73

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java (hunk 8 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: It looks like this case was an artifact of the way the tests were previously written, but is no longer needed. So I'm going to remove the extra test.

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java (hunk 19 above)
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java]: These tests never get as far as file comparison; they pass because the test using this data provider has an expectedException because ReadSparkSink currently only handles coordinate sorted output.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java (hunk 7 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Done.

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java (hunk 14 above)
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: No, it was created with samtools.  Still, I added another test using a queryname sorted cram file created with the new htsjdk code.Unfortunately in doing so I noticed some [other issues](https://github.com/samtools/htsjdk/issues/506) with the htsjdk CRAM code, but those are unrelated to this PR.
